homeassistant/components/tplink/light.py

```diff
@@ -48,31 +48,25 @@
 LIGHT_STATE_ERROR_MSG = "err_msg"
 
 LIGHT_SYSINFO_MAC = "mac"
+LIGHT_SYSINFO_HOST = "host"
 LIGHT_SYSINFO_ALIAS = "alias"
 LIGHT_SYSINFO_MODEL = "model"
 LIGHT_SYSINFO_IS_DIMMABLE = "is_dimmable"
 LIGHT_SYSINFO_IS_VARIABLE_COLOR_TEMP = "is_variable_color_temp"
 LIGHT_SYSINFO_IS_COLOR = "is_color"
 
+MAX_ATTEMPTS = 20
+SLEEP_TIME = 2
+
 
 async def async_setup_entry(hass: HomeAssistantType, config_entry, async_add_entities):
     """Set up switches."""
-    await async_add_entities_retry(
-        hass, async_add_entities, hass.data[TPLINK_DOMAIN][CONF_LIGHT], add_entity
-    )
+    for device in hass.data[TPLINK_DOMAIN][CONF_LIGHT]:
+        await hass.async_add_executor_job(device.get_sysinfo)
+        async_add_entities([TPLinkSmartBulb(device)], update_before_add=True)
```

Comment:
@TheGardenMonkey The below is untested but should do something close to what you want:

```suggestion
    devices = hass.data[TPLINK_DOMAIN][CONF_LIGHT]
    aws = [hass.async_add_executor_job(device.get_sysinfo) for device in devices]
    results = await asyncio.gather(*aws, return_exceptions=True)
    entities = []
    for idx, device in enumerate(devices):
    	if isinstance(results[idx], Exception):
    	    _LOGGER.warning("Failed to setup device: %s", device.name, exc_info=results[idx])
    	    continue
    	    
    	entities.append(TPLinkSmartBulb(device))
    	
    async_add_entities(entities, update_before_add=True)
```

Reply:
Maybe not submit all executor jobs concurrently to avoid claiming many threads at the same time?

Reply:
Thank you!!! I will test this shorty and update

Reply:
> Maybe not submit all executor jobs concurrently to avoid claiming many threads at the same time?

Missed this comment...is it reasonable to batch them (do x at a time)?

Reply:
@TheGardenMonkey 

If they have more than 64 devices we could overload the executor as @MartinHjelmare pointed out.

Something like this will work in that case.

```
    devices = hass.data[TPLINK_DOMAIN][CONF_LIGHT]
    entities = []
    for device in devices:
    	await hass.async_add_executor_job(device.get_sysinfo)
    	entities.append(TPLinkSmartBulb(device))
    	
    async_add_entities(entities, update_before_add=True)
```

I don't remember if device.get_sysinfo can throw.  If it can you might need something like....

```
    devices = hass.data[TPLINK_DOMAIN][CONF_LIGHT]
    entities = []
    for device in devices:
		try:
	    	await hass.async_add_executor_job(device.get_sysinfo)
		catch EXCPETION_HERE:
			???
			continue
    	entities.append(TPLinkSmartBulb(device))
    	
    async_add_entities(entities, update_before_add=True)
```

Reply:
Got it, will test this out now

Reply:
> > Maybe not submit all executor jobs concurrently to avoid claiming many threads at the same time?
> 
> 
> 
> Missed this comment...is it reasonable to batch them (do x at a time)?
> 
> 

It shouldn't be a problem to do 3-4 at a time if you are finding it's slow starting up

Reply:
Is there a need to return true in async_setup_entry? I have seen some integrations do it and others not. I don't see it referenced in the docs as being needed.

Reply:
The platform `async_setup_entry` should not return anything. The integration `async_setup_entry` should `return True`.

Reply:
> > > Maybe not submit all executor jobs concurrently to avoid claiming many threads at the same time?
> > 
> > 
> > Missed this comment...is it reasonable to batch them (do x at a time)?
> 
> It shouldn't be a problem to do 3-4 at a time if you are finding it's slow starting up

Tested, seems the fine, i don't see any difference in the started up time though. I didn't notice any cpu spikes or high wait times either. As a reference I have 30+ tplink devices.